java/arcs/android/storage/database/AndroidSqliteDatabaseManager.kt

```diff
@@ -82,59 +82,54 @@ class AndroidSqliteDatabaseManager(
                 }
+            }.also {
+                dbCache[entry.name to entry.isPersistent] = it
+            }
         }
     }
 
     override suspend fun snapshotStatistics(): Map<DatabaseIdentifier, Snapshot> = mutex.withLock {
         dbCache.mapValues { it.value.snapshotStatistics() }
     }
 
-    override suspend fun resetAll() {
-        registry.fetchAll()
-            .map { getDatabase(it.name, it.isPersistent) as DatabaseImpl }
-            .forEach { it.reset() }
+    override suspend fun resetAll() = runOnAllDatabases { _, db ->
+        db.reset()
     }
 
-    override suspend fun removeExpiredEntities(): Job = coroutineScope {
-        launch {
-            registry.fetchAll()
-                .map { it.name to getDatabase(it.name, it.isPersistent) }
-                .forEach { (name, db) ->
-                    if (databaseSizeTooLarge(name)) {
-                        // If the database size is too large, we clear it entirely.
-                        db.removeAllEntities()
-                    } else {
-                        db.removeExpiredEntities()
-                    }
-                }
+    override suspend fun removeExpiredEntities() = runOnAllDatabases { name, db ->
+        if (databaseSizeTooLarge(name)) {
+            // If the database size is too large, we clear it entirely.
+            db.removeAllEntities()
+        } else {
+            db.removeExpiredEntities()
         }
     }
 
-    override suspend fun removeAllEntities() =
-        registry.fetchAll()
-            .map { getDatabase(it.name, it.isPersistent) }
-            .forEach { it.removeAllEntities() }
+    override suspend fun removeAllEntities() = runOnAllDatabases { _, db ->
+        db.removeAllEntities()
+    }
 
-    override suspend fun removeEntitiesCreatedBetween(startTimeMillis: Long, endTimeMillis: Long) =
-        registry.fetchAll()
-            .map { getDatabase(it.name, it.isPersistent) }
-            .forEach { it.removeEntitiesCreatedBetween(startTimeMillis, endTimeMillis) }
-
-    override suspend fun runGarbageCollection(): Job = coroutineScope {
-        launch {
-            registry.fetchAll()
-                .map { getDatabase(it.name, it.isPersistent) }
-                .forEach { it.runGarbageCollection() }
-        }
+    override suspend fun removeEntitiesCreatedBetween(
+        startTimeMillis: Long,
+        endTimeMillis: Long
+    ) = runOnAllDatabases { _, db ->
+        db.removeEntitiesCreatedBetween(startTimeMillis, endTimeMillis)
+    }
+
+    override suspend fun runGarbageCollection() = runOnAllDatabases { _, db ->
+        db.runGarbageCollection()
     }
 
+    /** Execute the provided block on each of the registered databases. */
+    private suspend fun runOnAllDatabases(block: suspend (name: String, db: Database) -> Unit) =
+        registry.fetchAll()
+            .map { it.name to getDatabase(it.name, it.isPersistent) }
+            .forEach { block(it.first, it.second) }
```

Comment:
Since we currently only really have up to two databases (one in-memory and one persistent), it's not a big deal.. but in the future if/when there are more databases - we might consider converting the `forEach` into a `map` where the lambda launches a job for each item... then before returning - we just `joinAll`.

Reply:
I can do that in this CL. Would it be better to wrap the whole thing in a `coroutineScope`, rather than map/joinall?